experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,957 @@
+#include "dynamatic/Dialect/HW/HWDialect.h"
+#include "dynamatic/Dialect/HW/HWOpInterfaces.h"
+#include "dynamatic/Dialect/HW/HWOps.h"
+#include "dynamatic/Dialect/Handshake/HandshakeDialect.h"
+#include "dynamatic/Dialect/Handshake/HandshakeOps.h"
+#include "dynamatic/Dialect/Handshake/HandshakeTypes.h"
+#include "dynamatic/Support/RTL.h"
+#include "dynamatic/Support/System.h"
+#include "dynamatic/Support/TimingModels.h"
+#include "experimental/tools/handshake-simulator/ExecModels.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/OwningOpRef.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Support/IndentedOstream.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Support/LogicalResult.h"
+#include "llvm/ADT/Any.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Bitcode/LLVMBitCodes.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/InitLLVM.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/SMLoc.h"
+#include "llvm/Support/SourceMgr.h"
+#include "llvm/Support/raw_ostream.h"
+#include <algorithm>
+#include <any>
+#include <bitset>
+#include <cstdint>
+#include <iterator>
+#include <map>
+#include <memory>
+#include <optional>
+#include <set>
+#include <string>
+#include <system_error>
+#include <utility>
+#include <variant>
+#include <vector>
+
+using namespace llvm;
+using namespace mlir;
+using namespace dynamatic;
+using namespace handshake;
+
+static cl::OptionCategory mainCategory("Tool options");
+
+static cl::opt<std::string> inputFileName(cl::Positional,
+                                          cl::desc("<input file>"),
+                                          cl::init("-"), cl::cat(mainCategory));
+
+/// Abstract class for representing Wire. It may be signal (one boolean value),
+/// data (array) or whatever u want
+class Wire {
+public:
+  // check if wire has value (not undefined)
+  virtual bool hasValue() = 0;
+  // implement wire = <concrete_constant_value>
+  virtual void updateValue(std::any val) = 0;
+  // get the value of the wire
+  virtual std::any getData() = 0;
+  // print function for convenience
+  virtual void printValue() = 0;
+  virtual ~Wire() = default;
+};
+
+/// represent the concrete port as a smart pointer to the memory
+using Port = std::shared_ptr<Wire>;
```

Comment:
I really don't like this `using`. `Port` and `Wire` semantically seem to be completely different things, whereas the former is just a pointer of the latter.